crates/red_knot_python_semantic/src/types/call.rs

```diff
@@ -1,423 +1,262 @@
-                match error_kind {
-                    StaticAssertionErrorKind::ArgumentIsFalse => {
-                        context.report_lint(
-                            &STATIC_ASSERT_ERROR,
-                            node,
-                            format_args!("Static assertion error: argument evaluates to `False`"),
-                        );
-                    }
-                    StaticAssertionErrorKind::ArgumentIsFalsy(parameter_ty) => {
-                        context.report_lint(
-                            &STATIC_ASSERT_ERROR,
-                            node,
-                            format_args!(
-                                "Static assertion error: argument of type `{parameter_ty}` is statically known to be falsy",
-                                parameter_ty=parameter_ty.display(context.db())
-                            ),
-                        );
+                Err(error) => match error {
+                    CallError::NotCallable { not_callable_ty } => {
+                        not_callable.push(NotCallableVariant::new(not_callable_ty, None));
                     }
-                    StaticAssertionErrorKind::ArgumentTruthinessIsAmbiguous(parameter_ty) => {
-                        context.report_lint(
-                            &STATIC_ASSERT_ERROR,
-                            node,
-                            format_args!(
-                                "Static assertion error: argument of type `{parameter_ty}` has an ambiguous static truthiness",
-                                parameter_ty=parameter_ty.display(context.db())
-                            ),
-                        );
+                    CallError::NotCallableVariants {
+                        called_ty: _,
+                        callable: inner_callable,
+                        not_callable: inner_not_callable,
+                    } => {
+                        not_callable.extend(inner_not_callable);
+                        bindings.extend(inner_callable);
                     }
-                    StaticAssertionErrorKind::CustomError(message) => {
-                        context.report_lint(
-                            &STATIC_ASSERT_ERROR,
-                            node,
-                            format_args!("Static assertion error: {message}"),
-                        );
+                    // Should this be OK, or an error? We can't make it not_callable
+                    // because calling it would actually work because it ignores the
+                    // possibly unboundness.
+                    CallError::PossiblyUnboundDunderCall { outcome, .. } => match *outcome {
+                        CallOutcome::Union(inner_bindings) => {
+                            bindings.extend(inner_bindings);
+                        }
+                        CallOutcome::Single(binding) => {
+                            bindings.push(binding);
+                        }
+                    },
```

Comment:
I'm having trouble grappling with this question. Can you come up with a Python snippet where we'd change behaviour if we did this branch differently? It would help me think about the issue more clearly, I think

Reply:
I think the example here would be:

```py
def outer(cond1: bool):
	class Test:
		if cond:
			def __call__(self): ...
	
	class Other:
		def __call__(self): ...
	
	def inner(cond: bool): 
		if cond:
			a = Test()
		else:
			a = Other()
	
		a()
```

`a` resolves to `Test | Other`. Calling `Test` resolves to `PossiblyUnbound` and calling `Other` to `Ok` (let's ignore whether the `__call__` is actually correct). 

I do think what I have now is incorrect and I should push it to the error variant instead.

Reply:
I added a test case for this (with a TODO to improve the error message)